### PR TITLE
Shorten url: Unification across Explore and Dashboards

### PR DESCRIPTION
**Pull request by ivanahuckova**
**What this PR does / why we need it**:
We have created 2 PRs that add shortlink functionality to 
- Explore: https://github.com/grafana/grafana/pull/28222
- Dashboard: https://github.com/grafana/grafana/pull/27409

This PR unifies the methods and functions, so all shortlink generations and functionalities are in 1 place aka we are creating single source of truth. 

**Which issue(s) this PR fixes**:

This will check out 1 of the point in #28248

### CODE REVIEW

File: public/app/core/utils/shortLinks.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
+import memoizeOne from 'memoize-one';
+import { getBackendSrv, config } from '@grafana/runtime';
+import { AppEvents } from '@grafana/data';
+import appEvents from 'app/core/app_events';
+import { copyStringToClipboard } from './explore';
+
+function buildHostUrl() {
+  return `${window.location.protocol}//${window.location.host}${config.appSubUrl}`;
```
Comment (hugohaggmark) [inline public/app/core/utils/shortLinks.ts:8]: nit: Not sure this if it helps in this case, but I think there is an Url class that handles building urls.

File: public/app/core/utils/shortLinks.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+function getRelativeURLPath(url: string) {
+  let path = url.replace(buildHostUrl(), '');
+  return path.startsWith('/') ? path.substring(1, path.length) : path;
+}
```
Comment (hugohaggmark) [inline public/app/core/utils/shortLinks.ts:14]: nit: Same comment as above

**Review by hugohaggmark — APPROVED** [review]
I think this looks great ⭐   and I left some nits for consideration.

File: public/app/features/dashboard/components/ShareModal/ShareLink.tsx [hunk 3]
```diff
@@ -58,19 +58,18 @@ export class ShareLink extends PureComponent<Props, State> {
...
     if (useShortUrl) {
-      getBackendSrv()
-        .post(`/api/short-urls`, {
-          path: getRelativeURLPath(shareUrl),
-        })
-        .then(res => this.setState({ shareUrl: res.url, imageUrl }));
+      const shortLink = await createShortLink(shareUrl);
+      if (shortLink) {
+        this.setState({ shareUrl: shortLink, imageUrl });
+      }
```
Comment (hugohaggmark) [inline public/app/features/dashboard/components/ShareModal/ShareLink.tsx:72]: Hmm, this logic might be better in the `buildShareUrl` function, WDYT?

File: public/app/core/utils/shortLinks.ts (hunk 1 above)
Comment (ivanahuckova) [reply public/app/core/utils/shortLinks.ts:8]: I wasn't able to find Url/Link class where I could move this/use it from.

File: public/app/features/dashboard/components/ShareModal/ShareLink.tsx (hunk 3 above)
Comment (ivanahuckova) [reply public/app/features/dashboard/components/ShareModal/ShareLink.tsx]: Great point, I totally agree. I have moved `createShortLink` to `buildShareUrl`.

File: public/app/core/utils/shortLinks.ts (hunk 1 above)
Comment (hugohaggmark) [reply public/app/core/utils/shortLinks.ts:8]: https://developer.mozilla.org/en-US/docs/Web/API/URL

Comment (ivanahuckova) [reply public/app/core/utils/shortLinks.ts:8]: haha no worries! I thought that we have some class in Grafana that is handling this. But yeah, I looked into URL class and didn't find anything that would be helpful and solved any of our problems that we are solving.

**Pull request merged**